Cython/Compiler/TypeStubGenerator.py

```diff
@@ -0,0 +1,433 @@
+from . import Version
+from .Nodes import CNameDeclaratorNode
+from .ExprNodes import CallNode, NameNode, ImportNode, TupleNode, AttributeNode
+from ..CodeWriter import DeclarationWriter
+from .Visitor import CythonTransform
+from . import PyrexTypes
+from ..Utils import open_new_file
+import cython 
+import os 
+import sys 
+
+cython.declare(PyrexTypes=object, Naming=object, ExprNodes=object, Nodes=object,
+               Options=object, UtilNodes=object, LetNode=object,
+               LetRefNode=object, TreeFragment=object, EncodedString=object,
+               error=object, warning=object, copy=object, _unicode=object)
+
+
+# Inspired by and based around https://github.com/cython/cython/pull/3818
+# with some less lazy changes to it and a few minor improvements and optimizations...
+
+# Decided to revert to an older variant I had wrote of this code for the sake of 
+# maintainability - Vizonex
+
+
+
+# TODO Save this implementation commented out if required....
+if sys.version_info >= (3, 9):
+    typing_module = "typing"
+else:
+    typing_module = "typing_extensions"
+
+class PyiWriter(CythonTransform, DeclarationWriter):
```

Comment:
I don't think you use the directives tracking of `CythonTransform`. I wonder if this should inherit from `TreeVisitor` since I don't think it's intended to change the tree - just to go over it. That'd eliminate any possibility of accidentally changing the tree.

Reply:
I also don't know how much of `DeclarationWriter` you use - possibly just things like `write/indent`. I wonder if it would be worth factoring those out so you can inherit without bringing in all the `visit_` functions?